man/aarchup.1
.TH "aarchup" "1" "DEC 2018" "Leonidas Spyropoulos <artafinde AT gmail DOT com>" ""
.SH "NAME"
aarchup \- lightweight pacman update notifier, a fork from archup by Rorschach
.SH "SYNOPSIS"
\fIaarchup\fR [options]
.SH "DESCRIPTION"
aarchup is a small C++ application which informs the user when system\-updates for Archlinux are available. It's licensed under the GPLv3. In contrast to other update notifiers aarchup is intended to be lightweight and just do what it should, notify about possible updates.

aarchup uses GTk+ and libnotify to show a desktop notification if updates are available. It follows the unix-philosophy of "just doing one thing, but doing it well". It notifies about new updates. With a systemd timer aarchup can be used to regularly check for new updates and get a desktop notification if there are some. It can also runs on it's own and check for updates regularly.
.SH "OPTIONS"
          --command|-c [value]        Set the command which gives out the list of updates.
                                      The default is /usr/bin/checkupdates
          --icon|-p [value]           Shows the icon, whose path has been given as value, in the notification.
                                      By default no icon is shown.
          --maxentries|-m [value]     Set the maximum number of packages which shall be displayed in the notification.
                                      The default value is 30.
          --timeout|-t [value]        Set the timeout after which the notification disappears in seconds.
                                      The default value is 3600 seconds, which is 1 hour.
          --uid|-i [value]            Set the uid of the process.
                                      The default is to keep the uid of the user who started aarchup.
                                      !!You should change this if root is executing aarchup!!
          --urgency|-u [value]        Set the libnotify urgency-level. Possible values are: low, normal and critical.
                                      The default value is normal. With changing this value you can change the color of the notification.
          --loop-time|-l [value]      When this is used the program will control the check for updates in the interval of minutes specified,
                                      if none is specified, the default(60) will be used. See: Loop-time above.
          --help                      Prints this help.
          --version                   Shows the version.
          --aur                       Check aur for new packages too. Will need auracle installed.
          --debug|-d                  Print debug info.
          --ftimeout [value]          Program will manually enforce timeout for closing notification.
                                      Do NOT use with --timeout, if --timeout works or without --loop-time [value].
                                      The value for this option should be in minutes.
          --pkg-no-ignore             If this flag is set will not use the IgnorePkg variable from pacman.conf. Without the flag will ignore those
                                      packages. Feel free to add packages names to pacman.conf even if they are from AUR, will cause no harm and will
                                      also ignore it and not show updates.



.SH "CONFIGURATION"
aarchup can simply be invoked by executing it from the commandline. But you surely want to automate this task and let aarchup continuously be run with a systemd timer.
There are two ways of automating aarchup runs. One setup is to use the --loop-time option and the other is using a systemd timer.

\fILoop-time\fR

When using the --loop-time option the program will run endless. This has an advantage over the systemd method. For example on gnome3 when running aarchup with systemd, if you get more than one notification of updates and you don't close them, they will keep getting stacked and you are going to end up with a few notifications(of the same thing) at the notification bar. Which can get really annoying to close manually.
When the program is running on its own it can keep track of it's notifications and update them as needed instead of creating new ones.
In case you would like to use this method on startup copy /usr/share/doc/aarchup/aarchup.desktop to /home/user/.config/autostart

.PP
         $ sudo cp /usr/share/doc/aarchup/aarchup.desktop /home/user/.config/autostart/
.PP
where user is your username.
Copy the file /usr/share/doc/aarchup/aarchupstartup.example
.PP
         $ sudo cp /usr/share/doc/aarchup/aarchupstartup.example /usr/share/aarchup/aarchupstartup.sh
.PP
Edit the new file with different options if needed. Include --aur if you want to check for updates in AUR(you will need cower installed).

Aarchup will run on next boot or you can manually start it running /usr/share/aarchup/aarchupstartup.sh


\fIHourly\fR

The most simple setup is for a single user system where you want to check once an hour for updates. In this case just type:
.PP
          $ sudo systemctl start aarchup.timer
.PP
and make it enabled at boot:
.PP
          $ sudo systemctl enable aarchup.timer
.PP 
Now every hour your package database will get updated and after that aarchup will be executed. If there are updates aarchup shows a desktop notification, if there are no updates nothing will happen. The desktop notification will automatically disappear after 60min or if you simply click on it.

\fIother Intervals\fR

If you want to execute aarchup at other intervals than hourly, you can override the settings of the systemd timer unit by either ...
.PP
* Using the \fisystemctl --user edit $UNITNAME\fr command to interactively create an override file for the unit in \fi~/.config/systemd/user/$UNITNAME.d/\fr where $UNITNAME stands for the name of the systemd unit. 

* Or by creating an override file in the aforementioned directory, followed by executing \fisystemctl --user daemon-reload\fr
.PP
After you made the change, you need to restart the unit as well.

Some hints on this:
 * you can adjust the timeout value, before the notification will disappear with the --timeout option of aarchup

.SH "AUTHORS"
Written originally by Rorschach. aarchup by aericson. Current C++ version by Leonidas Spyropoulos.
.SH "CREDITS"
Most of the credits goes to Rorshach for the original archup.
The features made in the fork was made by aericson and can
be found in the CHANGELOG file.
Also some minor changes were made by Andrew Kravchuk and can be
found at his github(https://github.com/lockie/archup) logs.
.SH "REPORTING BUGS"
Bugs? If you find one please raise an Issue at GitHub https://github.com/inglor/aarchup
.SH "COPYRIGHT"
Copyright 2018 Leonidas Spyropoulos <artafinde AT gmail DOT com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. 


src/CHANGELOG
aarchup-2.1.0 ()
    * remove pkg-no-ignore param, fix some logic
aarchup-2.0.0 ()
    * rewritten in c++
aarchup-1.8 (01/10/2015)
    * fix to work with new systemd/dbus
aarchup-1.8 (30/09/2015)
    * Never run pacman -Sy, good bye partial upgrades (https://github.com/aericson/aarchup/issues/10)
    * remove --ignore-disconnect
aarchup-1.7.4 (11/08/2015)
    * Fix bug on noxaarchup.sh
aarchup-1.7.3 (12/02/2015)
    * Add missing .png file
aarchup-1.7.2 (05/02/2015)
    * updated noxaarchup.sh with another method to get user
aarchup-1.7.1 (01/02/2015)
    * updated noxaarchup.sh to work without gconf-helper
aarchup-1.7.0 (25/01/2015)
    * drop cron support
    * add systemd timer
    * update documentation
aarchup-1.6.8 (09/10/2013)
    * fix cronhourly.example
    * update documentation
aarchup-1.6.7 (24/02/2011)
    * typo on manpages
    * addded a .png icon to be used with WM that won't support svg
aarchup-1.6.6 (09/09/2011)
    * bug fixes - check commit changes
aarchup-1.6.5 (05/09/2011)
    * aarchupstartup.sh editted to kill running aarchup and restarting it(might fix some problems)
    * since pacman -Qu shows current version, added a '+' sign to the version name so that it won't be miss interpretated.
    * some small changes nothing really important
aarchup-1.6.4 (10/07/2011)
    * Bug fix(check commit)
aarchup-1.6.3 (09/07/2011)
    * Bug fixes (Fail to read /etc/pacman.conf when options like ShowSize is uncomment)
aarchup-1.6.2 (08/07/2011)
    * will read /etc/pacman.conf and won't notify for IgnorePkg updates
      You can add aur packages there too probably won't have any trouble.
      Please use IgnorePkg only once and separate packages with space
    * code refactoring
aarchup-1.6.1 (26/06/2011)
    * better debug messages can grep now using "info" or "error"
    * fixed buffer problem where it was not possible to redirect output to a file.
    * fixed bug regarding notification not being showed if it was showed before an enviroment restart.
aarchup-1.6 (22/06/2011)
    * wont notify about new updates when "pacman -Sy" fail, probably you don't have internet connection.
      You can change this behavior with --ignore-disconnect. Forcing the check for updates even after "pacman -Sy" fail.
    * better arguments parsing with getopt(GNU)
    * you can now optionaly use /etc/aarchup.conf instead of passing a bunch of arguments to the command-line.
      Check manpages for details.
aarchup-1.5.7 (04/06/2011)
    * added --ftimeout(with loop-time only) option, workaround for gnome not using --timeout.
    * will close notifications after next run if upgraded system and no notification was found(with loop-time only).
aarchup-1.5.6 (29/05/2011)
    * changed titles for notifications
    * added subdivision for regular updates and AUR updates
    more info: https://bbs.archlinux.org/viewtopic.php?pid=940866#p940866
aarchup-1.5.5 (22/05/2011)
    * bug fixes and edited cronhourly.example
aarchup-1.5.4 (22/05/2011)
    * fixed a bug that was delaying notifications until next run in --loop-time option. thanks lucke
aarchup-1.5.3 (21/05/2011)
    * added -d to print debug info
aarchup-1.5.2 (21/05/2011)
    * fixed typo in cronhourly.example
aarchup-1.5.1 (15/05/2011)
    * fixed memory allocation errors noted by rck at (https://bbs.archlinux.org/viewtopic.php?pid=875332#p875332)
aarchup-1.5 (15/05/2011)
    * added --loop-time flag to allow use without cronjob, check man pages for advantages
    * added default icon
    * added --aur flag to allow checking for updates in AUR. Will need cower installed.
### ORIGINAL ARCHUP ###
archup-1.4
    * added --icon|-p flag to show an icon in the notification
archup-1.3 (03/04/2010)
	* some code cleanup, updated docs, ...
	* added --uid|-i flag to allow dropping the rights if executed as root-cronjob
	* added --command|-c flag to allow a custom updatelist retrival command
	* added --urgency|-u flag to choose the urgencylevel and thus the color of the notification
archup-1.2 (21/03/2010)
	* improved sourcecode documentation
	* added commandline options (--help,--version,--timeout|-tm,--maxentries|-m)
	* moved timeout and maxentries values from compiletime-options to runtime-options
	* replaced the static memory managment with a dynamic one using realloc
	* added manpage
	* added $DESTDIR support in the make install and make uninstall commands
archup-1.1 (20/03/2010)
archup-1.0 (no public release)


src/aarchup.cpp
#include <ctype.h>
#include <getopt.h>
#include <libnotify/notify.h>
#include <plog/Appenders/ConsoleAppender.h>
#include <plog/Log.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <iostream>
#include <memory>
#include "CliWrapper.hh"

#define AUR_HEADER "AUR updates:\n"
#define VERSION_NUMBER "2.1.0"

/* Prints the help. */
int print_help() {
  std::cout
      << "Usage: aarchup [options]\n\n"
         "Options:\n"
         "          --command|-c [value]        Set the command which gives "
         "out the list of updates.\n"
         "                                      The default is "
         "/usr/bin/checkupdates\n"
         "          --icon|-p [value]           Shows the icon, whose path has "
         "been given as value, in the notification.\n"
         "                                      By default no icon is shown.\n"
         "          --maxentries|-m [value]     Set the maximum number of "
         "packages which shall be displayed in the notification.\n"
         "                                      The default value is 30.\n"
         "          --timeout|-t [value]        Set the timeout after which "
         "the notification disappears in seconds.\n"
         "                                      The default value is 3600 "
         "seconds, which is 1 hour.\n"
         "          --uid|-i [value]            Set the uid of the process.\n"
         "                                      The default is to keep the uid "
         "of the user who started aarchup.\n"
         "                                      !!You should change this if "
         "root is executing aarchup!!\n"
         "          --urgency|-u [value]        Set the libnotify "
         "urgency-level. Possible values are: low, normal and critical.\n"
         "                                      The default value is normal. "
         "With changing this value you can change the color of the "
         "notification.\n"
         "          --loop-time|-l [value]      Using this the program will "
         "loop indefinitely. Just use this if you won't use cron.\n"
         "                                      The value should be the number "
         "of minutes from each update check, if none is informed 60 will be "
         "used.\n"
         "                                      For more information on it "
         "check man.\n"
         "          --help|-h                   Prints this help.\n"
         "          --version|-v                Shows the version.\n"
         "          --aur                       Check aur for new packages "
         "too. Will need auracle installed.\n"
         "          --debug|-d                  Print debug info.\n"
         "          --ftimeout|-f [value]       Program will manually enforce "
         "timeout for closing notification.\n"
         "                                      Do NOT use with --timeout, if "
         "--timeout works or without --loop-time [value].\n"
         "                                      The value for this option "
         "should be in minutes.\n"
         "\nMore information can be found in the manpage.\n";
  exit(0);
}

/* Prints the version. */
int print_version() {
  std::cout
      << "aarchup " VERSION_NUMBER
         "\nCopyright 2018 Leonidas Spyropoulos <artafinde@gmail.com>\n"
         "License GPLv3+: GNU GPL version 3 or later "
         "<http://gnu.org/licenses/gpl.html>.\n"
         "This is free software, you are free to change and redistribute it.\n"
         "There is NO WARRANTY, to the extent permitted by law.\n";
  exit(0);
}

std::vector<std::string> split(const std::string &s, char delimiter) {
  std::vector<std::string> tokens;
  std::string token;
  std::istringstream tokenStream(s);
  while (std::getline(tokenStream, token, delimiter)) {
    tokens.push_back(token);
  }
  return tokens;
}

int main(int argc, char **argv) {
  NotifyUrgency urgency = NOTIFY_URGENCY_NORMAL;
  const char *command = "/usr/bin/checkupdates";
  const char *aurCommand = "/usr/bin/auracle sync";

  long timeout = 3600 * 1000;
  long max_number_out = 30;
  long loop_time = 3600;
  long manual_timeout = 0;
  gchar *icon = nullptr;
  bool will_loop = FALSE;
  static int help_flag = 0;
  static int version_flag = 0;
  static int aur = 0;

  plog::ConsoleAppender<plog::TxtFormatter> consoleAppender;
  plog::init(plog::warning, &consoleAppender);

  if (argc > 1) {
    if (strcmp(argv[1], "--version") == 0 || strcmp(argv[1], "-v") == 0)
      print_version();
    if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-h") == 0)
      print_help();
  }

  /* Parse commandline options */
  const char *const short_opts = "c:p:m:t:i:u:l:df:";
  const option long_opts[] = {
      {"command", required_argument, nullptr, 'c'},
      {"icon", required_argument, nullptr, 'p'},
      {"maxentries", required_argument, nullptr, 'm'},
      {"timeout", required_argument, nullptr, 't'},
      {"uid", required_argument, nullptr, 'i'},
      {"urgency", required_argument, nullptr, 'u'},
      {"loop-time", required_argument, nullptr, 'l'},
      {"help", no_argument, &help_flag, 1},
      {"version", no_argument, &version_flag, 1},
      {"aur", no_argument, &aur, 1},
      {"ftimeout", required_argument, nullptr, 'f'},
      {"debug", no_argument, nullptr, 'd'},
      {nullptr, 0, nullptr, 0},
  };

  while (true) {
    int option_index = 0;
    const auto opt =
        getopt_long(argc, argv, short_opts, long_opts, &option_index);
    if (-1 == opt) {
      break;
    }
    /* Short opts */
    switch (opt) {
      case 'd':
        plog::get()->setMaxSeverity(plog::verbose);
        break;
      case 0:
        if (long_opts[option_index].flag) {
          break;
        }
        LOGE << "Undefined option " << long_opts[option_index].name;
        if (optarg) {
          LOGE << " with args " << optarg;
        }
        break;
      case 'v':
        LOGV << "Printing version";
        print_version();
        break;
      case 'c':
        command = optarg;
        LOGV << "Command set: '" << command << "'";
        break;
      case 'p':
        icon = optarg;
        LOGV << "Icon set: '" << icon << "'";
        break;
      case 'm':
        if (!isdigit(optarg[0])) {
          LOGF << "Argument '--maxentries' should be number";
          exit(1);
        }
        max_number_out = std::stol(optarg);
        LOGV << "Max_number set: '" << max_number_out << "' lines";
        break;
      case 't':
        if (!isdigit(optarg[0])) {
          LOGF << "Argument '--timeout' should be number";
          exit(1);
        }
        timeout = std::stol(optarg) * 1000;
        LOGV << "Timeout set: " << timeout / 1000 << " sec(s)";
        ;
        break;
      case 'i':
        if (!isdigit(optarg[0])) {
          LOGF << "Argument '--uid' should be number";
          exit(1);
        }
        if (setuid(static_cast<__uid_t>(std::stol(optarg))) != 0) {
          LOGF << "Couldn't change to the given uid, aborting";
          exit(1);
        }
        LOGV << "Setting uid to: " << optarg;
        break;
      case 'u':
        if (strcmp(optarg, "low") == 0) {
          urgency = NOTIFY_URGENCY_LOW;
        } else if (strcmp(optarg, "normal") == 0) {
          urgency = NOTIFY_URGENCY_NORMAL;
        } else if (strcmp(optarg, "critical") == 0) {
          urgency = NOTIFY_URGENCY_CRITICAL;
        } else {
          LOGF << "Argument '--urgency' has to be 'low', 'normal' or 'critical";
          exit(1);
        }
        LOGV << "Urgency set: " << urgency;
        break;
      case 'l':
        will_loop = TRUE;
        if (!isdigit(optarg[0])) {
          LOGF << "Argument '--loop-time' should be number";
          exit(1);
        }
        loop_time = std::stol(optarg) * 60;
        LOGV << "Loop_time set: " << loop_time / 60 << " min(s)";
        break;
      case 'f':
        if (!isdigit(optarg[0])) {
          LOGF << "Argument '--ftimeout' should be number";
          exit(1);
        }
        manual_timeout = std::stol(optarg) * 60;
        if (!will_loop) {
          LOGF << "Argument '--ftimeout' can't be used without or before "
                  "'--loop-time'";
          exit(1);
        }
        if (manual_timeout > loop_time) {
          LOGF << "Please set a value for '--ftimeout' that is lower than "
                  "'--loop-time'";
          exit(1);
        }
        LOGV << "Manual_timeout: " << manual_timeout / 60 << " min(s)";
        break;
      case 'h':
      case '?':
        print_help();
        break;
      default:
        exit(1);
    }
  }

  long offset = 0;
  NotifyNotification *my_notify = nullptr;
  const char *name = "New Updates";
  const char *category = "update";
  GError *error = nullptr;
  do {
    LOGD << "Executing command '" << command << "' for updates";
    auto checkUpdatesCmd = std::make_unique<CliWrapper>(command);
    const std::string &checkUpdateOut = checkUpdatesCmd->execute();
    std::string aurHelperOut;
    if (aur) {
      LOGD << "Executing command '" << aurCommand << "' for AUR updates";
      auto aurHelperCmd = std::make_unique<CliWrapper>(aurCommand);
      aurHelperOut = aurHelperCmd->execute();
    }
    if (!checkUpdateOut.empty() || (!aurHelperOut.empty())) {
      std::string finalOut = "There are updates for:\n";
      if (aurHelperOut.empty()) {
        finalOut = finalOut + checkUpdateOut;
      } else if (checkUpdateOut.empty()) {
        finalOut = finalOut + AUR_HEADER + aurHelperOut;
      } else {
        finalOut = finalOut + checkUpdateOut + AUR_HEADER + aurHelperOut;
      }
      auto outputLines = split(finalOut, '\n');
      int lines = 0;
      std::stringstream ss;
      for (auto &outputLine : outputLines) {
        ss << outputLine << '\n';
        lines++;
        if (lines >= max_number_out) {
          break;
        }
      }
      if (!notify_is_initted()) {
        notify_init(name);
      }
      bool persist = TRUE;
      gboolean success;
      do {
        if (!my_notify) {
          my_notify = notify_notification_new(
              "New updates for Arch Linux available!", ss.str().c_str(), icon);
        } else {
          notify_notification_update(my_notify,
                                     "New updates for Arch Linux available!",
                                     ss.str().c_str(), icon);
        }
        notify_notification_set_timeout(my_notify, timeout);
        notify_notification_set_category(my_notify, category);
        notify_notification_set_urgency(my_notify, urgency);
        success = notify_notification_show(my_notify, &error);
        if (success)
          LOGD << "Notification shown successfully";
        else {
          if (persist) {
            LOGW << "Notification failed, reason:\n\t[" << error->code << "] "
                 << error->message << "\n";
          } else {
            LOGE << "Notification failed, reason:\n\t[" << error->code << "] "
                 << error->message << "\n";
          }
          g_error_free(error);
          error = nullptr;
          if (persist) {
            LOGW << "It could have been caused by an environment restart, "
                    "trying to work around it by re-init libnotify";
            my_notify = nullptr;
            notify_uninit();
            notify_init(name);
            persist = FALSE;
          }
        }
      } while (!my_notify);
      if (error) {
        g_error_free(error);
        error = nullptr;
      }
      if (manual_timeout && success && will_loop) {
        LOGD << "Will close notification in " << manual_timeout / 60
             << " minutes (this time will be reduced from the loop-time)";
        sleep(static_cast<unsigned int>(manual_timeout));
        offset = manual_timeout;
        if (notify_notification_close(my_notify, &error))
          LOGD << "Notification closed";
        else {
          LOGW << "Failed to close, reason:\n\t[" << error->code << "] "
               << error->message;
        }
        if (error) {
          g_error_free(error);
          error = nullptr;
        }
      }
    } else {
      LOGI << "No updates found";
      if (my_notify) {
        LOGD << "Previous notification found. Closing it in case it was "
                "still opened";
        if (notify_notification_close(my_notify, &error))
          LOGD << "Notification closed";
        else {
          LOGW << "Failed to close, reason:\n\t[" << error->code << "] "
               << error->message;
        }
        if (error) {
          g_error_free(error);
          error = nullptr;
        }
      }
    }

    if (will_loop) {
      LOGD << "Next run will be in " << (loop_time - offset) / 60 << " minutes";
      sleep(static_cast<unsigned int>(loop_time - offset));
      offset = 0;
    }
  } while (will_loop);
  return 0;
}
